src/uct/ib/base/ib_iface.c

```diff
@@ -828,42 +828,39 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
         if ((config->pkey == UCS_HEXUNITS_AUTO) ||
             /* take only the lower 15 bits for the comparison */
             ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
-            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK)) {
-                if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
-                    /* limited PKEY has not yet been found */
-                    lim_pkey_index = pkey_index;
-                    lim_pkey       = pkey;
-                }
-                continue;
+            if (pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) {
+                iface->pkey_index = pkey_index;
+                iface->pkey       = pkey;
+                pkey_found        = 1;
+                goto out_pkey_found;
+            } else if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
+                /* limited PKEY has not yet been found */
+                lim_pkey_index = pkey_index;
+                lim_pkey       = pkey;
             }
-
-            iface->pkey_index = pkey_index;
-            iface->pkey       = pkey;
-            pkey_found        = 1;
-            break;
         }
     }
 
-    if (!pkey_found) {
-        if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
-            /* PKEY neither with full nor with limited membership was found */
-            if (config->pkey == UCS_HEXUNITS_AUTO) {
-                ucs_error("there is no valid pkey to use on "
-                          UCT_IB_IFACE_FMT, UCT_IB_IFACE_ARG(iface));
-            } else {
-                ucs_error("unable to find specified pkey 0x%x on "UCT_IB_IFACE_FMT,
-                          config->pkey, UCT_IB_IFACE_ARG(iface));
-            }
+    ucs_assert(!pkey_found);
 
-            return UCS_ERR_NO_ELEM;
+    if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) {
+        /* PKEY neither with full nor with limited membership was found */
+        if (config->pkey == UCS_HEXUNITS_AUTO) {
+            ucs_error("there is no valid pkey to use on " UCT_IB_IFACE_FMT,
+                      UCT_IB_IFACE_ARG(iface));
         } else {
-            ucs_assert(lim_pkey_index != UINT16_MAX);
-            iface->pkey_index = lim_pkey_index;
-            iface->pkey       = lim_pkey;
+            ucs_error("unable to find specified pkey 0x%x on "UCT_IB_IFACE_FMT,
+                      config->pkey, UCT_IB_IFACE_ARG(iface));
         }
+
+        return UCS_ERR_NO_ELEM;
+    } else {
+        ucs_assert(lim_pkey_index != UINT16_MAX);
```

Comment:
don't we already check `if (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY)`

Reply:
it is to make sure that we don't miss to set `lim_pkey_index`